feat: add access control for session recordings

zlwaterfield opened the pull request:
## Changes

Introduces session recording as a resource type for access control throughout the frontend and backend. Updates UI  and API to respect user access levels for session recording actions.

Session Recordings are unique compared to existing models. 
- a user does not create them 
- they can't be "edited" other than by deleting

Because of the above, there are no "object access controls", only "resource access controls".

This PR introduces 2 new patterns:
- model access control inheritance
- field access controls

**Model access control inheritance**
Using a map, a model can inherit the access controls from another model. Basically mocking the access controls for it. In this PR Session Recording Playlist inherit the resource access controls from Session Recordings. We do this because Session Recording Playlists are so tightly correlated with Session Recordings, we want to apply the same permissions. So if a user has view-only access to the Session Recordings resource, then they also have view-only access to Session Recording Playlists. 

**Field access controls**
There are many cases where a model can be affected by writes to another model, so this addresses that issue. In this case, the settings for Session Recordings like `session_recording_opt_in` (which means if it's enabled or not) are on the Team model. We want to ensure that if someone has view-only access to Session Recordings, they can't change the settings for it. This pattern will apply to a lot of models so I've created a simpe mapper to help.

Note: this does not add the client-side checks for the new interface @adamleithp is building because I want to chat with him about how the HOC structure should work. It's an unreleased feature and API still blocks the changes so shouldn't be an issue. 

## How did you test this code?

Added tests for the two new patterns.

Tested manually a lot.

zlwaterfield committed:
Add access control for session recordings

Introduces session recording as a resource type for access control throughout the frontend and backend. Updates UI components to respect user access levels for session recording actions, and adds backend enforcement and tests for session recording permissions.

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

zlwaterfield committed:
Add access control to session recording playlist actions

Replaced LemonButton with AccessControlledLemonButton for creating, duplicating, and pinning session recording playlists, enforcing minimum access level requirements. Updated PinToPlaylistButton to use AccessControlAction for access checks. These changes ensure only users with appropriate permissions can perform playlist-related actions.

zlwaterfield committed:
Add access control to pin playlist button

Replaced LemonButton with AccessControlledLemonButton for the pin/unpin action in saved session recording playlists. The button now requires Editor access level for session recordings, enforcing resource-based access control.

zlwaterfield committed:
Add resource and field-level access control for session recording playlists

Introduces RESOURCE_INHERITANCE_MAP and FIELD_ACCESS_CONTROL_MAP to support access inheritance and field-level permissions, specifically for session_recording_playlist and Team model fields. Updates UserAccessControl and UserAccessControlSerializerMixin to enforce these rules. Adds comprehensive tests for resource inheritance and field-level access control. Integrates access control mixins into session_recording_playlist API and serializer.

zlwaterfield committed:
Remove comments

benjackwhite commented on frontend/src/scenes/session-recordings/saved-playlists/SavedSessionRecordingPlaylists.tsx:
```diff
@@ -154,10 +162,15 @@ export function SavedSessionRecordingPlaylists({ tab }: SavedSessionRecordingPla
             render: function Render(pinned, { short_id }) {
                 return (
-                    <LemonButton
+                    <AccessControlledLemonButton
```
I think at some point we should consider just making these optional properties on the standard LemonButton

zlwaterfield replied:
https://github.com/PostHog/posthog/pull/37452 working on a PR here for it

benjackwhite commented on posthog/rbac/user_access_control.py:
```diff
@@ -54,8 +54,42 @@ class AccessSource(Enum):
+# Field-level access control mapping
+# Format: {ModelName: {field_name: (resource, required_level)}}
+FIELD_ACCESS_CONTROL_MAP: dict[str, dict[str, tuple[APIScopeObject, AccessControlLevel]]] = {
+    "Team": {
```
It would be _so sweet_ if we could do this on the model itself tbh. Like as a decorator or something. That way the thinking about the setting is at the model and not in a separate file (like when we add it in the future and forget to modify this separate file)

pauldambra reviewed (approved):
couple of thoughts but only being a noise box

got stuck trying to sort out the billing environment variables to test it
but it works in the case where there is no access control at least 😅 

so i'm happy for you to carry on without me testing it specifically

zlwaterfield commented on posthog/session_recordings/test/test_session_recording_access_control.py:
```diff
@@ -0,0 +1,194 @@
+            organization_member=membership,
+        )
+
+    def test_viewer_can_retrieve_recording(self):
```
Sure i can test that.

zlwaterfield committed:
Add field-level access control to Team model

Introduces a field_access_control decorator for specifying access requirements on model fields. Applies this decorator to session recording fields in the Team model, replacing static FIELD_ACCESS_CONTROL_MAP with dynamic metadata extraction. Updates related tests and serializer mixin to use the new dynamic field access control mapping.

zlwaterfield committed:
Remove TYPE_CHECKING guard from imports

Imports for AccessControlLevel and APIScopeObject are now unconditional, ensuring they are always available at runtime.

a-lider reviewed (changes requested):
I'm having circular import issues locally: `ImportError: cannot import name 'Organization' from partially initialized module 'posthog.models' (most likely due to a circular import)`

The cycle looks like this:
session_recording.py imports Team -> team imports field_access_control from rbac.decorators -> decorators.py imports AccessControlLevel -> user_access_control.py imports Team

zlwaterfield committed:
Refactor field_access_control to function and update usage

Changed field_access_control from a decorator to a helper function that directly attaches access control metadata to model fields. Updated all usages in Team model and related tests to use the new function signature for improved clarity and maintainability.

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/rbac/sesion-replay

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

zlwaterfield committed:
Move imports under TYPE_CHECKING in decorators.py

Imports for AccessControlLevel and APIScopeObject are now only imported during type checking to avoid runtime import overhead and potential circular dependencies.

zlwaterfield committed:
fix test types

zlwaterfield committed:
Fix tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

a-lider reviewed (approved):
Looks solid! Tested the main flow, everything worked. 

As an idea, we could also add the `Access control` button in the side panel of session recordings, similar to Dashboards and Insights.

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/rbac/sesion-replay

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/rbac/sesion-replay

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)
